### PR TITLE
p2p/simulations: Snapshot support

### PR DESCRIPTION
**Pull request by lmars**
This adds the concept of a simulation network snapshot which contains all the nodes and connections which exist in the network as well as the state of each node.

Node services optionally implement a `Snapshot` method which returns arbitrary data, and then that snapshot can be passed back when starting the node in another network.

This removes the need for a journal (which was intended as a way to restore a network from a previous simulation run), so I have removed the journal code.

I've also updated `SimNode` so that it initializes a new service each time `Start` is called (rather than just starting the same service every time).

### CODE REVIEW

File: p2p/simulations/events.go [hunk 1]
```diff
@@ -28,13 +28,16 @@ func NewEvent(v interface{}) *Event {
 	switch v := v.(type) {
 	case *Node:
 		event.Type = EventTypeNode
-		event.Node = v
+		node := *v
```
[the remark this answers is not in the document]
Comment (lmars) [reply p2p/simulations/events.go:57]: We need to copy the object so the event represents the state of the object at this time (not copying means the object can be mutated in the future and the event is then different, e.g. `node.Up=true`).

I have added a comment: https://github.com/ethersphere/go-ethereum/blob/d7eb6b5a70d4f49887ce982068007821dad8b407/p2p/simulations/events.go#L50-L51

File: p2p/simulations/http.go [hunk 2]
```diff
@@ -260,6 +272,8 @@ func NewServer(config *ServerConfig) *Server {
 	s.GET("/networks", s.GetNetworks)
 	s.GET("/networks/:netid", s.GetNetwork)
 	s.GET("/networks/:netid/events", s.StreamNetworkEvents)
+	s.POST("/networks/:netid/snapshots/create", s.CreateSnapshot)
```
[the remark this answers is not in the document]
Comment (lmars) [reply p2p/simulations/http.go:275]: > creating a snapshot should be a GET /networks/:netid/snaphot i mean why POST?

A `GET` request should never be "creating" things, we are creating a new snapshot here not getting an existing one, hence `POST`.

> loading should be PUT /networks/:netid/snapshot why POST and 'create' is never part of the url no?

`PUT` is supposed to be "store this resource at this URL", we aren't doing that here, we are mutating the state of the network by essentially providing a diff, so I think `POST` is more appropriate, and also makes it easier to perform from a web browser (e.g. just "upload" a snapshot).

File: p2p/simulations/http.go [hunk 3]
```diff
@@ -381,6 +395,37 @@ func (s *Server) StreamNetworkEvents(w http.ResponseWriter, req *http.Request) {
...
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	if err := network.Load(snap); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	s.JSON(w, http.StatusOK, network)
```
[the remark this answers is not in the document]
Comment (lmars) [reply p2p/simulations/http.go:426]: We have loaded a snapshot into the network and produced a new state of the network, so here we are returning that new network state.

File: p2p/simulations/http_test.go [hunk 4]
```diff
@@ -345,3 +384,129 @@ func TestHTTPNodeRPC(t *testing.T) {
...
+		}
+		if string(state) != states[i] {
+			t.Fatalf("expected snapshot state %q, got %q", states[i], state)
+		}
+	}
+
+	// check we got all the events
+	x := &expectEvents{t, events, sub}
+	x.expect(
+		x.nodeEvent(nodes[0].ID, false),
```
[the remark this answers is not in the document]
Comment (lmars) [reply p2p/simulations/http_test.go:505]: When loading a snapshot, we create and start each node one by one, and then restore the connections, and that is represented by the order here.

File: p2p/simulations/network.go [hunk 5]
```diff
@@ -571,3 +576,69 @@ func (self *Network) Shutdown() {
...
+
+func (self *Network) Snapshot() (*Snapshot, error) {
+	self.lock.Lock()
+	defer self.lock.Unlock()
+	snap := &Snapshot{
+		Nodes: make([]NodeSnapshot, len(self.Nodes)),
+		Conns: make([]Conn, len(self.Conns)),
+	}
+	for i, node := range self.Nodes {
+		snapshot, err := node.Snapshot()
```
[the remark this answers is not in the document]
Comment (lmars) [reply p2p/simulations/network.go]: I have modified this to only snapshot running nodes: https://github.com/ethersphere/go-ethereum/blob/d7eb6b5a70d4f49887ce982068007821dad8b407/p2p/simulations/network.go#L615-L625

File: p2p/simulations/adapters/exec.go [hunk 6]
```diff
@@ -71,11 +69,11 @@ func (e *ExecAdapter) NewNode(config *NodeConfig) (Node, error) {
...
 	node := &ExecNode{
-		ID:      config.Id,
-		Service: config.Service,
-		Dir:     dir,
-		Config:  &conf,
-		key:     config.PrivateKey,
+		ID:          config.Id,
+		Service:     config.Service,
+		Dir:         dir,
+		StackConfig: &conf,
```
[the remark this answers is not in the document]
Comment (lmars) [reply p2p/simulations/adapters/exec.go]: I have reduced the separation by setting `ExecNode.Config` to `*execNodeConfig` which wraps both the stack and node config, better?

https://github.com/ethersphere/go-ethereum/blob/d7eb6b5a70d4f49887ce982068007821dad8b407/p2p/simulations/adapters/exec.go#L93

File: p2p/simulations/adapters/inproc.go [hunk 7]
```diff
@@ -161,39 +154,60 @@ func (self *SimNode) ServeRPC(conn net.Conn) error {
...
 	}
-	return self.service.Start(self)
+	self.running = service
+	return service.Start(&simServer{self})
+}
+
+// simServer wraps a SimNode but modifies the Start method signature so that
+// it implements the p2p.Server interface (the Start method is never actually
+// called when using the SimAdapter)
+type simServer struct {
```
[the remark this answers is not in the document]
Comment (lmars) [reply p2p/simulations/adapters/inproc.go:180]: The sim node needs to implement both the `adapters.Node` interface and the `p2p.Server` interface, but they have conflicting `Start` method signatures, so we wrap `SimNode` in `simServer` to implement `p2p.Server.Start`, leaving `SimNode` implementing `adapters.Node.Start`.

Makes sense?

File: p2p/simulations/http.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zelig) [reply p2p/simulations/http.go:275]: Cmon man, when I search google, I create a search right? no the resources there are networks, nodes and connections, we dont create or change anything there. A snapshot is actually the same as an update or index. 
In fact it does not even change any state on the network so it is a very clear case for GET

for load, i disagree as well, what you quote for PUT is the purist PUT, everybody uses PUT as PATCH basically, ie., update partial info on an existing resource. or change its state.
That is what we are doing.

POST a snapshot is also possible but it should mean that the network itself is created in the very same step. But on second thought that is probably the more common usecase but then the action is not `loadSnapshot` but `CreateNetworkFromSnapshot`

**Pull request merged**